rust/arrow/src/array/union.rs

```diff
@@ -0,0 +1,1174 @@
+        assert!((type_id as usize) < self.boxed_fields.len());
+        self.boxed_fields[type_id as usize].clone()
+    }
+
+    /// Returns the `type_id` for the array slot at `index`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index` is greater than the length of the array.
+    pub fn type_id(&self, index: usize) -> i8 {
+        assert!(index - self.offset() < self.len());
+        self.data().buffers()[0].data()[index] as i8
+    }
+
+    /// Returns the offset into the underlying values array for the array slot at `index`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index` is greater than the length of the array.
+    pub fn value_offset(&self, index: usize) -> i32 {
+        assert!(index - self.offset() < self.len());
+        if self.is_dense() {
+            let valid_slots = match self.data.null_buffer() {
+                Some(b) => bit_util::count_set_bits_offset(b.data(), 0, index),
+                None => index,
+            };
+            self.data().buffers()[1].data()[valid_slots * size_of::<i32>()] as i32
+        } else {
+            index as i32
+        }
+    }
+
+    /// Returns the array's value at `index`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index` is greater than the length of the array.
+    pub fn value(&self, index: usize) -> ArrayRef {
+        let type_id = self.type_id(self.offset() + index);
+        let value_offset = self.value_offset(self.offset() + index) as usize;
+        let child_data = self.boxed_fields[type_id as usize].clone();
+        child_data.slice(value_offset, 1)
+    }
+
+    /// Returns the names of the types in the union.
+    pub fn type_names(&self) -> Vec<&str> {
+        match self.data.data_type() {
+            DataType::Union(fields) => fields
+                .iter()
+                .map(|f| f.name().as_str())
+                .collect::<Vec<&str>>(),
+            _ => unreachable!("Union array's data type is not a union!"),
+        }
+    }
+
+    /// Returns whether the `UnionArray` is dense (or sparse if `false`).
+    fn is_dense(&self) -> bool {
+        self.data().buffers().len() == 2
+    }
+}
+
+impl From<ArrayDataRef> for UnionArray {
+    fn from(data: ArrayDataRef) -> Self {
+        let mut boxed_fields = vec![];
+        for cd in data.child_data() {
+            boxed_fields.push(make_array(cd.clone()));
+        }
+        Self { data, boxed_fields }
+    }
+}
+
+impl Array for UnionArray {
+    fn as_any(&self) -> &Any {
+        self
+    }
+
+    fn data(&self) -> ArrayDataRef {
+        self.data.clone()
+    }
+
+    fn data_ref(&self) -> &ArrayDataRef {
+        &self.data
+    }
+}
+
+impl fmt::Debug for UnionArray {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let header = if self.is_dense() {
+            "UnionArray(Dense)\n[\n"
+        } else {
+            "UnionArray(Sparse)\n[\n"
+        };
+        write!(f, "{}", header)?;
+
+        write!(f, "-- type id buffer:\n")?;
```

Comment:
There's a recent PR that fixed some `clippy` lints, I think `writeln!(f, "...")` is preferable over `write!(f, "...\n")`

Reply:
Yep, will update.